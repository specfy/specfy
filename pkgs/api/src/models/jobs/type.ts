import type { Prisma } from '@prisma/client';

export interface JobDeployConfig {
  url: string;
  autoLayout?: boolean;
  hook?: {
    id: string;
    ref: string;
  };
}

export type JobWithOrgProject = Prisma.JobsGetPayload<{
  include: { Org: true; Project: true };
}>;

export type JobCode =
  | 'cant_auth_github'
  | 'failed_to_checkout'
  | 'failed_to_cleanup'
  | 'failed_to_clone'
  | 'failed_to_deploy'
  | 'failed_to_start_github_deployment'
  | 'failed_to_teardown'
  | 'missing_dependencies'
  | 'no_api_key'
  | 'org_not_installed'
  | 'project_not_installed'
  | 'success'
  | 'unknown';
