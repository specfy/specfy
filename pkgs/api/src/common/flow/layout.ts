import type { ComputedFlow, Layout, Tree } from './types.js';

const paddingY = 20;
const paddingX = 20;
const paddingHost = 20;

/**
 * Compute a nested tree representing Flow.nodes .
 */
export function computeTree(
  nodes: ComputedFlow['nodes'],
  parentId: string | null = null
): Tree[] {
  const childs: Tree[] = [];
  for (const node of nodes) {
    if (parentId && node.parentNode !== parentId) {
      continue;
    }
    if (!parentId && node.parentNode) {
      continue;
    }

    childs.push({
      id: node.id,
      parentId,
      childs: computeTree(nodes, node.id),
    });
  }

  return childs;
}

/**
 * Depth first algorithm.
 *
 * To compute a layout we:
 * - Build a tree that will allow us to compute Bounding box of an "Host"
 * - We go at the deepest level, when there is no more host available
 * - We compute place the components and compute the BB
 * - Once we have the BB, we go back one level
 * - At this level, we know exactly how big the childs are so we can also compute the BB + padding
 * - etc. until level 0
 */
export function computeLayout(flow: ComputedFlow): Layout {
  const trees = computeTree(flow.nodes);

  const layout = computeTreeLayout(trees, flow.nodes);

  return layout;
}

export function computeTreeLayout(
  trees: Tree[],
  nodes: ComputedFlow['nodes']
): Layout {
  let y = 20;
  let w = 0;
  const x = paddingHost;
  const subs: Array<{ id: string; layout: Layout }> = [];

  // Compute aggregate (host) first
  for (const tree of trees) {
    if (tree.childs.length <= 0) {
      continue;
    }
    subs.push({ id: tree.id, layout: computeTreeLayout(tree.childs, nodes) });
  }

  const layout: Layout = { nodes: [], height: 0, width: 0, y: 0, x: 0 };

  // Place host first on the layout
  for (const sub of subs) {
    layout.nodes.push(...sub.layout.nodes);

    const node = nodes.find((n) => n.id === sub.id)!;
    layout.nodes.push({
      id: node.id,
      pos: { x, y },
      size: { width: sub.layout.width, height: sub.layout.height },
    });

    // increase the BB
    w = Math.max(w, x + sub.layout.width + paddingX);
    y += sub.layout.height + paddingY;
  }

  // Then place the components
  for (const tree of trees) {
    if (tree.childs.length > 0) {
      continue;
    }

    const node = nodes.find((n) => n.id === tree.id)!;
    layout.nodes.push({
      id: node.id,
      pos: { x, y },
      size: node.data.originalSize,
    });

    // increase the BB
    y += node.data.originalSize.height + paddingY;
    w = Math.max(w, x + node.data.originalSize.width + paddingX);
  }

  // Update the Bounding Box
  layout.height = y;
  layout.width = w;

  return layout;
}

/**
 * Apply the computed layout to the Flow.
 * This will modify the reference.
 */
export function applyLayout(layout: Layout, flow: ComputedFlow) {
  for (const node of flow.nodes) {
    const rel = layout.nodes.find((l) => l.id === node.id)!;
    node.position = rel.pos;
    node.style = {
      width: `${rel.size.width}px`,
      height: `${rel.size.height}px`,
    };
  }
}
