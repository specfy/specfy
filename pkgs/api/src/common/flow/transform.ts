import type { Edge, Node } from 'reactflow';

import type {
  ComponentForFlow,
  ComputedFlow,
  EdgeData,
  NodeData,
} from './types.js';

export function createNode(component: ComponentForFlow): Node<NodeData> {
  const node: Node<NodeData> = {
    id: component.id,
    type: 'custom',
    data: {
      label: component.name,
      type: component.type,
      techId: component.techId,
      originalSize: component.display.size,
    },
    position: { ...component.display.pos },
    targetPosition: 'left' as any,
    sourcePosition: 'right' as any,
    style: {
      width: `${component.display.size.width}px`,
      height: `${component.display.size.height}px`,
    },
    deletable: true,
    connectable: true,
    draggable: true,
    focusable: true,
    selectable: true,
  };

  if (component.inComponent) {
    node.extent = 'parent';
    node.parentNode = component.inComponent;
  }

  return node;
}

export function getEdgeMarkers(data: EdgeData) {
  const edge: Partial<Edge> = {};
  if (data.read) {
    edge.markerStart = {
      type: 'arrowclosed' as any,
      width: 10,
      height: 10,
    };
  }
  if (data.write) {
    edge.markerEnd = {
      type: 'arrowclosed' as any,
      width: 10,
      height: 10,
    };
  }
  return edge;
}
export function componentsToFlow(components: ComponentForFlow[]): ComputedFlow {
  const edges: Array<Edge<EdgeData>> = [];
  const nodes: Array<Node<NodeData>> = [];

  // Create all hosting nodes
  // We need to add them first because React Flow is not reordering
  const hosts = components.filter((comp) => comp.type === 'hosting');
  const done: string[] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (hosts.length <= 0) {
      break;
    }
    const host = hosts.shift()!;
    if (host.inComponent && !done.includes(host.inComponent)) {
      hosts.push(host);
      continue;
    }

    nodes.push(createNode(host));
    done.push(host.id);
  }

  // Create all other nodes
  for (const comp of components) {
    if (comp.type === 'hosting') {
      continue;
    }

    nodes.push(createNode(comp));
  }

  for (const comp of components) {
    for (const edge of comp.edges) {
      const item: Edge<EdgeData> = {
        id: `${comp.id}->${edge.to}`,
        source: comp.id,
        target: edge.to,
        sourceHandle: edge.portSource,
        targetHandle: edge.portTarget,
        data: { read: edge.read, write: edge.write },
        ...getEdgeMarkers(edge),
        // type: 'floating',
      };
      edges.push(item);
    }
  }

  return { edges, nodes };
}
