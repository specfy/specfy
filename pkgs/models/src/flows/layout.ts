import { hDefHost, wDefHost } from './constants.js';
import type { ComputedFlow, Layout, Tree } from './types.js';

const paddingY = 30;
const paddingX = 30;
const paddingCompX = 100;
const paddingHost = 30;
const hostsPerRow = 2;
const compsPerRow = 2;

/**
 * Compute a nested tree representing Flow.nodes .
 */
export function computeTree(
  nodes: ComputedFlow['nodes'],
  parentId: string | null = null
): Tree[] {
  const childs: Tree[] = [];
  for (const node of nodes) {
    if (parentId && node.parentNode !== parentId) {
      continue;
    }
    if (!parentId && node.parentNode) {
      continue;
    }

    childs.push({
      id: node.id,
      parentId,
      childs: computeTree(nodes, node.id),
      isHost: node.data.type === 'hosting',
    });
  }

  return childs;
}

/**
 * Depth first algorithm.
 *
 * To compute a layout we:
 * - Build a tree that will allow us to compute Bounding box of an "Host"
 * - We go at the deepest level, when there is no more host available
 * - We compute place the components and compute the BB
 * - Once we have the BB, we go back one level
 * - At this level, we know exactly how big the childs are so we can also compute the BB + padding
 * - etc. until level 0
 */
export function computeLayout(flow: ComputedFlow): Layout {
  const trees = computeTree(flow.nodes);

  const layout = computeTreeLayout(trees, flow.nodes);

  return layout;
}

export function computeTreeLayout(
  trees: Tree[],
  nodes: ComputedFlow['nodes']
): Layout {
  let y = 20;
  let w = 0;
  let h = 0;
  let x = paddingHost;
  const subs: Array<{ id: string; layout: Layout }> = [];

  // Compute aggregate (host) first
  for (const tree of trees) {
    if (!tree.isHost) {
      continue;
    }
    subs.push({ id: tree.id, layout: computeTreeLayout(tree.childs, nodes) });
  }

  const layout: Layout = { nodes: [], height: 0, width: 0, y: 0, x: 0 };

  // Place host first on the layout
  for (let index = 0; index < subs.length; index++) {
    const sub = subs[index];
    if (index > 0 && index % hostsPerRow === 0) {
      // Go back to the start of the line
      y = h;
      x = paddingHost;
    } else if (index > 0) {
      // Add on the right of the previous element
      const prev = layout.nodes[layout.nodes.length - 1];
      x = prev.pos.x + prev.size.width + paddingX;
    }

    layout.nodes.push(...sub.layout.nodes);

    const node = nodes.find((n) => n.id === sub.id)!;
    layout.nodes.push({
      id: node.id,
      pos: { x, y },
      size: { width: sub.layout.width, height: sub.layout.height },
    });

    // increase the BB
    w = Math.max(w, x + sub.layout.width + paddingX);
    h = Math.max(h, y + sub.layout.height + paddingY);
  }

  // Then place the components
  for (let index = 0; index < trees.length; index++) {
    const tree = trees[index];
    if (tree.isHost) {
      continue;
    }

    if (index > 0 && index % compsPerRow === 0) {
      // Go back to the start of the line
      y = h;
      x = paddingHost;
    } else if (index > 0) {
      // Add on the right of the previous element
      const prev = layout.nodes[layout.nodes.length - 1];
      x = prev.pos.x + prev.size.width + paddingCompX;
    }

    const node = nodes.find((n) => n.id === tree.id)!;
    layout.nodes.push({
      id: node.id,
      pos: { x, y },
      size: node.data.originalSize,
    });

    // increase the BB
    h = Math.max(h, y + node.data.originalSize.height + paddingY);
    w = Math.max(w, x + node.data.originalSize.width + paddingCompX);
  }

  // Update the Bounding Box
  layout.height = Math.max(hDefHost, h);
  layout.width = Math.max(wDefHost, w);

  return layout;
}

/**
 * Apply the computed layout to the Flow.
 * This will modify the reference.
 */
export function applyLayout(layout: Layout, flow: ComputedFlow) {
  for (const node of flow.nodes) {
    const rel = layout.nodes.find((l) => l.id === node.id)!;
    node.position = rel.pos;
    node.style = {
      width: `${rel.size.width}px`,
      height: `${rel.size.height}px`,
    };
  }
}
