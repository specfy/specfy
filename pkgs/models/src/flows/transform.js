export function createNodeFromProject(project, display) {
    return createNode({
        id: project.id,
        name: project.name,
        type: 'project',
        display: display,
        inComponent: null,
        techId: null,
        typeId: null,
    });
}
export function createNode(component) {
    const node = {
        id: component.id,
        type: 'custom',
        data: {
            name: component.name,
            type: component.type,
            techId: component.techId,
            typeId: component.typeId,
            originalSize: component.display.size,
        },
        position: { ...component.display.pos },
        style: {
            width: `${component.display.size.width}px`,
            height: `${component.display.size.height}px`,
        },
    };
    if (component.inComponent) {
        node.extent = 'parent';
        node.parentNode = component.inComponent;
    }
    return node;
}
export function getEdgeMarkers(data) {
    const edge = {};
    // if (data.read) {
    //   edge.markerStart = {
    //     type: 'arrowclosed' as any,
    //     width: 10,
    //     height: 10,
    //   };
    // }
    if (data.write) {
        edge.markerEnd = {
            type: 'arrowclosed',
            width: 10,
            height: 10,
        };
    }
    return edge;
}
export function componentsToFlow(components) {
    const edges = [];
    const nodes = [];
    // Create all hosting nodes
    // We need to add them first because React Flow is not reordering
    const hosts = components.filter((comp) => comp.type === 'hosting');
    const done = [];
    let i = 0;
    while (i < 9999) {
        i += 1;
        if (hosts.length <= 0) {
            break;
        }
        const host = hosts.shift();
        if (host.inComponent && !done.includes(host.inComponent)) {
            hosts.push(host);
            continue;
        }
        nodes.push(createNode(host));
        done.push(host.id);
    }
    if (i >= 9999) {
        throw new Error("Can't compute host");
    }
    // Create all other nodes
    for (const comp of components) {
        if (comp.type === 'hosting') {
            continue;
        }
        nodes.push(createNode(comp));
    }
    for (const comp of components) {
        for (const edge of comp.edges) {
            const item = {
                id: `${comp.id}->${edge.target}`,
                source: comp.id,
                target: edge.target,
                sourceHandle: edge.portSource,
                targetHandle: edge.portTarget,
                data: { read: edge.read, write: edge.write },
                ...getEdgeMarkers(edge),
                // type: 'floating',
            };
            edges.push(item);
        }
    }
    return { edges, nodes };
}
