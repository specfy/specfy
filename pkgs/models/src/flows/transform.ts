import type { Projects } from '@specfy/db';

import type { DBComponent } from '../components/types.js';

import type {
  FlowItemDisplay,
  ComponentForFlow,
  ComputedEdge,
  ComputedFlow,
  ComputedNode,
  EdgeData,
} from './types.js';

export function createNodeFromProject(
  project: Pick<Projects, 'id' | 'name'>,
  display: FlowItemDisplay
) {
  return createNode({
    id: project.id,
    name: project.name,
    type: 'project',
    display: display,
    inComponent: { id: null },
    techId: null,
    typeId: null,
    source: null,
    show: true,
  });
}

export function createNode(
  component: Omit<ComponentForFlow, 'edges'>
): ComputedNode {
  const node: ComputedNode = {
    id: component.id,
    type: 'custom',
    data: {
      name: component.name,
      type: component.type,
      techId: component.techId,
      typeId: component.typeId,
      originalSize: component.display.size,
      source: component.source,
    },
    position: component.display.pos,
    style: {
      width: component.display.size.width,
      height: component.display.size.height,
    },
    width: component.display.size.width,
    height: component.display.size.height,
    hidden: component.show === false,
  };

  if (component.inComponent.id) {
    // node.extent = 'parent';
    node.parentNode = component.inComponent.id;
    node.expandParent = true;
  }

  return node;
}

export function getEdgeMarkers(data: EdgeData) {
  const edge: Partial<ComputedEdge> = {};
  // if (data.read) {
  //   edge.markerStart = {
  //     type: 'arrowclosed' as any,
  //     width: 10,
  //     height: 10,
  //   };
  // }
  if (data.write) {
    edge.markerEnd = {
      type: 'arrowclosed' as any,
      width: 14,
      height: 14,
    };
  }
  return edge;
}

export function componentsToFlow(components: ComponentForFlow[]): ComputedFlow {
  const edges: ComputedEdge[] = [];
  const nodes: ComputedNode[] = [];

  // Create all hosting nodes
  // We need to add them first because React Flow is not reordering
  const hosts = components.filter((comp) => comp.type === 'hosting');
  const done: string[] = [];

  let i = 0;
  while (i < 999) {
    i += 1;
    if (hosts.length <= 0) {
      break;
    }
    const host = hosts.shift()!;
    if (host.inComponent.id && !done.includes(host.inComponent.id)) {
      hosts.push(host);
      continue;
    }

    nodes.push(createNode(host));
    done.push(host.id);
  }
  if (i >= 998) {
    throw new Error("Can't compute host");
  }

  // Create all other nodes
  for (const comp of components) {
    if (comp.type === 'hosting') {
      continue;
    }

    nodes.push(createNode(comp));
  }

  for (const comp of components) {
    for (const edge of comp.edges) {
      const item: ComputedEdge = {
        id: `${comp.id}->${edge.target}`,
        source: comp.id,
        target: edge.target,
        sourceHandle: edge.portSource,
        targetHandle: edge.portTarget,
        data: { read: edge.read, write: edge.write, source: edge.source },
        hidden: edge.show === false,
        ...getEdgeMarkers(edge),
        // type: 'floating',
      };
      edges.push(item);
    }
  }

  return { edges, nodes };
}

export function getAbsolutePosition(
  component: DBComponent,
  components: DBComponent[]
) {
  let x = 0;
  let y = 0;
  let host: DBComponent | undefined = component;

  do {
    if (host) {
      x += host.display.pos.x;
      y += host.display.pos.y;

      host = host.inComponent.id
        ? components.find((c) => c.id === host?.inComponent.id)
        : undefined;
    }
  } while (host);

  return { x, y };
}

export function placeInsideHost(
  component: DBComponent,
  idHost: string,
  components: DBComponent[]
) {
  const host = components.find((c) => c.id === idHost);
  if (!host) {
    return { x: 0, y: 0 };
  }

  const absComp = getAbsolutePosition(component, components);
  const absHost = getAbsolutePosition(host, components);

  const x = Math.abs(absHost.x - absComp.x);
  const y = Math.abs(absHost.y - absComp.y);

  return { x, y };
}
