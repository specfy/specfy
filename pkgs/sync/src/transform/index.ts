import path from 'node:path';

import type { BaseProvider, ProviderFile } from '../provider/base.js';

export type TransformedFile = ProviderFile & { content: any };

export async function transform(
  provider: BaseProvider,
  files: ProviderFile[]
): Promise<TransformedFile[]> {
  const list = [];

  for (const file of files) {
    const content = await provider.open(path.join(provider.basePath, file.fp));

    list.push({ ...file, content });
  }

  return list;
}
