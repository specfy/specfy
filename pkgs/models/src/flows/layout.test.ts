import { describe, expect, it } from 'vitest';

import { computeLayout, computeTree } from './layout.js';
import { createNode } from './transform.js';
import type { ComponentForFlow } from './types.js';

function getComp(id: string, host: string | null = null): ComponentForFlow {
  return {
    id,
    edges: [],
    inComponent: host,
    type: 'service',
    typeId: null,
    name: id,
    techId: null,
    display: { pos: { x: 0, y: 0 }, size: { width: 100, height: 20 } },
  };
}

function getHost(id: string, host: string | null = null): ComponentForFlow {
  return {
    ...getComp(id, host),
    type: 'hosting',
  };
}

describe('layout', () => {
  it('should output 3 components', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [
        createNode(getComp('a')),
        createNode(getComp('b')),
        createNode(getComp('c')),
      ],
    });

    /**
     *      **A**   **B**
     *      **C**
     */
    expect(layout).toMatchSnapshot();
  });

  it('should compute a tree', () => {
    const tree = computeTree([
      createNode(getHost('a')),
      createNode(getHost('b', 'a')),
      createNode(getHost('c', 'b')),
      createNode(getComp('d', 'c')),
      createNode(getComp('e', 'a')),
      createNode(getComp('f')),
    ]);

    expect(tree).toMatchSnapshot();
  });

  it('should compute a layout', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [
        createNode(getHost('a')),
        createNode(getHost('b', 'a')),
        createNode(getHost('c', 'b')),
        createNode(getComp('d', 'c')),
        createNode(getComp('e', 'a')),
        createNode(getComp('f')),
      ],
    });

    /**
         +-----------------------------------------+
         |                                       A |   -----------
         | +-----------------------+   -------     |   f
         | |                     B |   e           |
         | |  +---------------+    |               |
         | |  | ------      C |    |               |
         | |  | d             |    |               |
         | |  +---------------+    |               |
         | |                       |               |
         | +-----------------------+               |
         +-----------------------------------------+
     */
    expect(layout).toMatchSnapshot();
  });

  it('should put 4 hosts in columns', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [
        createNode(getHost('a')),
        createNode(getHost('b')),
        createNode(getHost('c')),
        createNode(getHost('d')),
      ],
    });

    /**
          +----------+   +----------+
          |A         |   |B         |
          +----------+   +----------+

          +----------+   +----------+
          |C         |   |D         |
          +----------+   +----------+
     */
    expect(layout).toMatchSnapshot();
  });

  it('should put components in columns', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [
        createNode(getComp('a')),
        createNode(getComp('b')),
        createNode(getComp('c')),
        createNode(getComp('d')),
        createNode(getComp('e')),
      ],
    });

    /**
           +----------+   +----------+
           |    A     |   |    B     |
           +----------+   +----------+
           +----------+   +----------+
           |    C     |   |    D     |
           +----------+   +----------+
           +----------+
           |    E     |
           +----------+
     */
    expect(layout).toMatchSnapshot();
  });

  it('should put host inside host', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [createNode(getHost('a')), createNode(getHost('b', 'a'))],
    });

    expect(layout).toMatchSnapshot();
  });

  it('should put component inside host', () => {
    const layout = computeLayout({
      edges: [],
      nodes: [
        createNode(getHost('a')),
        createNode(getHost('b', 'a')),
        createNode(getComp('c', 'b')),
        createNode(getComp('d', 'a')),
        createNode(getComp('f')),
      ],
    });

    expect(layout).toMatchSnapshot();
  });
});
