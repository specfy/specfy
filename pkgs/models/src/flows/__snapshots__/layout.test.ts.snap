// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout > should compute a layout 1`] = `
{
  "height": 222,
  "nodes": [
    {
      "id": "d",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "c",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 72,
        "width": 230,
      },
    },
    {
      "id": "b",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 122,
        "width": 290,
      },
    },
    {
      "id": "e",
      "pos": {
        "x": 420,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 172,
        "width": 620,
      },
    },
    {
      "id": "f",
      "pos": {
        "x": 750,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
  ],
  "width": 950,
  "x": 0,
  "y": 0,
}
`;

exports[`layout > should compute a tree 1`] = `
[
  {
    "childs": [
      {
        "childs": [
          {
            "childs": [
              {
                "childs": [],
                "id": "d",
                "isHost": false,
                "parentId": "c",
              },
            ],
            "id": "c",
            "isHost": true,
            "parentId": "b",
          },
        ],
        "id": "b",
        "isHost": true,
        "parentId": "a",
      },
      {
        "childs": [],
        "id": "e",
        "isHost": false,
        "parentId": "a",
      },
    ],
    "id": "a",
    "isHost": true,
    "parentId": null,
  },
  {
    "childs": [],
    "id": "f",
    "isHost": false,
    "parentId": null,
  },
]
`;

exports[`layout > should output 3 components 1`] = `
{
  "height": 120,
  "nodes": [
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "b",
      "pos": {
        "x": 230,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "c",
      "pos": {
        "x": 30,
        "y": 70,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
  ],
  "width": 430,
  "x": 0,
  "y": 0,
}
`;

exports[`layout > should put 4 hosts in columns 1`] = `
{
  "height": 224,
  "nodes": [
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 72,
        "width": 140,
      },
    },
    {
      "id": "b",
      "pos": {
        "x": 200,
        "y": 20,
      },
      "size": {
        "height": 72,
        "width": 140,
      },
    },
    {
      "id": "c",
      "pos": {
        "x": 30,
        "y": 122,
      },
      "size": {
        "height": 72,
        "width": 140,
      },
    },
    {
      "id": "d",
      "pos": {
        "x": 200,
        "y": 122,
      },
      "size": {
        "height": 72,
        "width": 140,
      },
    },
  ],
  "width": 370,
  "x": 0,
  "y": 0,
}
`;

exports[`layout > should put component inside host 1`] = `
{
  "height": 172,
  "nodes": [
    {
      "id": "c",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "b",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 72,
        "width": 230,
      },
    },
    {
      "id": "d",
      "pos": {
        "x": 360,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 122,
        "width": 560,
      },
    },
    {
      "id": "f",
      "pos": {
        "x": 690,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
  ],
  "width": 890,
  "x": 0,
  "y": 0,
}
`;

exports[`layout > should put components in columns 1`] = `
{
  "height": 170,
  "nodes": [
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "b",
      "pos": {
        "x": 230,
        "y": 20,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "c",
      "pos": {
        "x": 30,
        "y": 70,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "d",
      "pos": {
        "x": 230,
        "y": 70,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
    {
      "id": "e",
      "pos": {
        "x": 30,
        "y": 120,
      },
      "size": {
        "height": 20,
        "width": 100,
      },
    },
  ],
  "width": 430,
  "x": 0,
  "y": 0,
}
`;

exports[`layout > should put host inside host 1`] = `
{
  "height": 172,
  "nodes": [
    {
      "id": "b",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 72,
        "width": 140,
      },
    },
    {
      "id": "a",
      "pos": {
        "x": 30,
        "y": 20,
      },
      "size": {
        "height": 122,
        "width": 200,
      },
    },
  ],
  "width": 260,
  "x": 0,
  "y": 0,
}
`;
